src/DS/sds.c

```diff
@@ -218,26 +218,50 @@ static struct oscap_source *load_referenced_source(const struct ds_sds_session *
 	return source_file;
 }
 
+static int ds_sds_register_sce(struct ds_sds_session *session, xmlNodePtr component_inner_root, const char* component_id, const char* target_filename_dirname, const char* relative_filepath)
+{
+	// the cast is safe to do because we are using the GNU basename, it doesn't
+	// modify the string
+	const char* file_basename = basename((char*)relative_filepath);
+	const char* sce_filename = oscap_sprintf("%s/%s/%s",ds_sds_session_get_target_dir(session), target_filename_dirname, file_basename);
+	int ret = ds_sds_dump_component_sce(component_inner_root->children, component_id, sce_filename);
+	oscap_free(sce_filename);
+	return ret;
+}
 
-static int ds_sds_dump_component(const char* external_file, const char* component_id, struct ds_sds_session *session, const char *target_filename_dirname, const char *relative_filepath)
+static int ds_sds_register_xmlDoc(struct ds_sds_session *session, xmlDoc* doc, xmlNodePtr component_inner_root, const char *relative_filepath)
 {
-	int ret = 0;
-	xmlDoc *doc;
-	struct oscap_source* source_file = NULL;
+	xmlDoc *new_doc = ds_doc_from_foreign_node(component_inner_root, doc);
+	if (new_doc == NULL) {
+		return -1;
+	}
 
-	if (external_file != NULL) {
+	struct oscap_source *component_source = oscap_source_new_from_xmlDoc(new_doc, relative_filepath);
 
-		source_file = load_referenced_source(session, external_file);
-		doc = oscap_source_get_xmlDoc(source_file);
+	return ds_sds_session_register_component_source(session, relative_filepath, component_source);
+}
 
-		if (doc == NULL) {
-			ret = -1;
-			goto cleanup;
-		}
+static int ds_sds_register_component(struct ds_sds_session *session, xmlDoc* doc, xmlNodePtr component_inner_root, const char* component_id, const char* target_filename_dirname, const char* relative_filepath)
+{
+	if (component_inner_root == NULL)
+	{
+		oscap_seterr(OSCAP_EFAMILY_XML, "Found component (id='%s') but it has no element contents, nothing to dump, skipping...", component_id);
+		return -1;
+	}
+
+	// If the inner root is script, we have to treat it in a special way
+	if (strcmp((const char*)component_inner_root->name, "script") == 0) {
+		return ds_sds_register_sce(session, component_inner_root, component_id, target_filename_dirname, relative_filepath);
+	// Otherwise we create a new XML doc we will dump the contents to.
+	// We can't just dump node "innerXML" because namespaces have to be
+	// handled.
```

Comment:
Comment is for "else", not for return. It is like here https://github.com/OpenSCAP/openscap/blob/maint-1.2/src/DS/sds.c#L275
I could move the comment after "else".